man/man1/xtfs_remove_osd.1
.TH xtfs_remove_osd 1 "July 2011" "The XtreemFS Distributed File System" "XtreemFS Admin Tools"
.SH NAME
xtfs_remove_osd \- removes an OSD by copying all objects on this OSD to other OSDs
.SH SYNOPSIS
\fBxtfs_remove_osd [ \fIoptions\fB ] \fIuuid:<osd_uuid>
.BR

.SH DESCRIPTION
.I xtfs_remove_osd
removes an OSD which is currently used. It copies all objectes on the OSD with the UUID 
.I <osd_uuid>
to other OSDs and then shut down the OSD. 

.SH EXAMPLE USAGE
.TP
.B "xtfs_remove_osd -dir localhost:32638 uuid:d1feb333-d986-4e91-b8e7-eb2a48485f8b"
Starts the removing process on the OSD identified by UUID 'd1feb333-d986-4e91-b8e7-eb2a48485f8b' and copies all files automatically to other OSDs (determinded by the OSD Selection Policy).

.SH OPTIONS
.TP
\fB-c \fI<creds_file>
Path to a PKCS#12 credentials file (private key + certificate) to use for SSL authentication. Must be present when MRC URL starts with pbrpcs:// or pbrpcg://.
.TP
\fB-cpass \fI<creds_passphrase>
An optional passphrase to access the credentials file.
.TP
\fB-dir \fI<uri>
Specifies the directory service to use (e.g. 'pbrpc://localhost:32638'). If no URI is specified, URI and security settings are taken from '/etc/xos/xtreemfs/default_dir'. In case of a secured URI ('pbrpcs://...' or 'pbrpcg://...'), it is necessary to also specify SSL settings.
.TP
\fB-h/--help
Shows usage info.
.TP
\fB-t \fI<trusted_CAs>
Path to a PKCS#12 file containing a set of certificates from trusted certification authorities. These certificates will be used to authenticate the MRC. Must be present when MRC URL starts with pbrpcs:// or pbrpcg://.
.TP
\fB-tpass \fI<trusted_passphrase>
An optional passphrase to access the truststore file.
.TP
\fB-d 
Enables debugging output. 
.TP
\fB-s
Shuts down the OSD per remote call after all object files are moved to other OSDs. Note that the shutdown process will be performed independently from your init system which can lead to errors if you try to start this OSD again with your init scripts. Therefore the default is not to shutdown the OSD. This step should be performed manually. 
.TP
.B \--admin_password \fI<passphrase>
The administrator password, if password protection is enabled. Used only if you want to shutdown the OSD automatically per remote call.
.TP
\fB---ssl-protocol \fI<ssl-protocol> 
SSL/TLS version to use: ssltls, tlsv1, tlsv11, tlsv12. 'ssltls' (default) accepts all versions, the others accept only the exact version they name. 'tlsv12' is available in JDK 7+ only. 'tlsv11' comes with JDK 6 or 7, depending on the vendor.


.SH LIMITATIONS
.TP
While the draining is in progress, it is not possible to modify the to be moved files as they are temporarily set to read-only. Additionally, when the draining will be started, no files of the OSD should be hold open by any client.
.TP
Draining an OSD that holds striped and replicated files involves certain limitations. A striped file is currently moved to a new OSD in its entirety rather than partially. Regardless of the original number of OSDs it was striped across, it is relocated to a single target OSD, which implies that the target file is not striped anymore. As a consequence, data may be moved from OSDs that are not directly affected by the draining process.
.TP
Furthermore, at least one OSD has to exist for each file that does not hold any data of the file, i.e. is not included in any stripe or replica. If all OSDs that are eligible according to the OSD selection policy have already been assigned to a file, draining any of these OSDs will fail, as no additional replicas can be created.

.SH "SEE ALSO"
.BR xtfs_chstatus (1),
.BR xtfs_mrcdbtool (1),
.BR xtfs_remove_osd (1),
.BR xtfs_scrub (1),
.BR xtfs_cleanup (1)
.BR

.SH AVAILABILITY
The xtfs_remove_osd command is part of the XtreemFS-tools package and is available at \fIhttp://www.xtreemfs.org\fP.


CHANGELOG
26-OCT-2016: release 1.6.0 (Phantastic Pancakes)

  Changes:
    * Extended volume quota implementation to OSDs. This requires dumping and reimporting the MRC database during upgrade (see 'MRC Database Backups and Conversion' in the user guide).
      * Implemented hard quotas for volumes and user and groups per volume.
      * Added quota information to import and export function.
      * These changes supersede the previous implementation.
    * Fix for mount.xtreemfs to fully support mount options via /etc/fstab.
    * Introduced Java wrappers for the C++ version of libxtreemfs and made it the default Java implementation.
    * Improved Hadoop adapter to work with a broader range of systems and versions.
    * Added 'listen.port.bind_retries' server option that implements exponential backoff during service startup if the port is not free yet.
    * Added RabbitMQ OSD tracing policy (ID 6003).
    * Added Preferred UUID OSD selection policy (ID 1003).
    * Extended X509 authentication to allow user and group mapping.
    * Slimmed down and updated Mac OS X installer with stricter dependency checking and SIP awareness.
    * Converted the Java parts to Maven and made all components available as artifacts hosted on GitHub.
    * Cleanup of code base to exclude generated sources and previously shipped binaries.
    * Add non-interactive console support to command line tools.
    * Remove explicit support for SSLv3.
    * Many small stability improvements and fixes.

12-MAR-2015: release 1.5.1

  Changes:
    * Hadoop Adapter supports Hadoop-2.x and other applications running on the YARN platform (tested with Apache Spark and Apache Flink), new OSD selection policy that imitates the HDFS placement behavior.
    * Consistent adding and removing replicas for R/W replication.
    * xtfs_scrup now replaces failed replicas using the R/W replication policy.
    * Improvements of the SSL mode: the used SSL/TLS version is selectable, strict certificate chain checks are possible, the SSL code on client and server side was improved.
    * Support for PKCS12 certificate containers is remove from the Windows client (did not work in previous release).
    * mount.xtreemfs accepts all parameters to be passed in the form -o option=value. If XtreemFS is mounted via /etc/fstab, mount options are passed to the XtreemFS client.
    * Dockerfiles for the XtreemFS services are available at https://github.com/xtreemfs/xtreemfs-docker.
    * A Vagrantfile is available in the XtreemFS repository to ease development and testing for new contributors.
    * Fixed snapshots in the case of deleted files.
    * Snapshots can be deleted even after inactivating snapshots for a volume (issue 290).
    * Initial version of an LD_PRELOAD based client that bypasses FUSE (experimental).
    * Initial implementation of volume quotas (checked only while opening files by the MRC).
    * OSDs can report their health to the DIR (e.g. based on SMART values), which aggregates this information in the web interface. A new OSD selection policy sorts out unhealthy OSDs.
    * Improved automatic testing: introduced Travis-CI, migrated all JUnit tests to version 4, improved reliability of nightly integration tests.
    * Added new benchmark tool xtfs_benchmark that is based on the JAVA libxtreemfs.
    * Many small improvements and fixes.

12-MAR-2014: release 1.5 (Wonderful Waffles)

  Changes:
    * Improved Hadoop Adapter e.g., added support for multiple volumes and read and write buffer to speed up performance.
    * Support for SSDs: Support for multiple OSD storage threads to increase parallelism.
    * Status webpage as part of the DIR webinterface for replicated files to show current primary and backup replicas.
    * Multi-homing support. XtreemFS can now be made available for multiple networks and clients will pick the correct address automatically.
    * Support for multiple OSDs per machine (e.g. one for each disk) through the xtreemfs-osd-farm init.d script.
    * Fixed major issues in Read-Only and Read/Write file replication.
    * Reduced metadata cache default timeout from 120 to 10 seconds. Disable it completely with --metadata-cache-size=0.
    * Upgraded to Google Protobuffers 2.5 which improves performance.
    * Windows Client: Fixed several issues (e.g., 286, 299) and updated the CbFS driver to the latest version.
    * XtreemFS also compiles under buildroot now.
    * Fixed several issues around SSL (e.g., 263, 281).
    * Fixed many small issues e.g., reported volume capacity was inaccurate (issue 296).
      Other fixed issues: 267, 268, 272, 274, 276, 277, 278, 283, 287, 288, 292, 293, 294.
    * Improved general stability (e.g., see issues 297, 301).
    * xtfs_scrub now repairs replicas with an invalid checksum automatically.
    * Disabled deferred deletion of file data on the OSDs.
    * Improved test coverage of nightly tests and added missing unit tests and a Valgrind leak check for C++ unit tests.

12-NOV-2012: release 1.4 (Salty Sticks)

  Changes:
    * Improved stability: Fixed client crashes when timeouts occurred; fixed server crashes and issues with the R/W replication.
    * Full support for asynchronous writes: If a client is started with "--enable-async-writes", write() requests will be immediately acknowledged by the client and executed in the background. Outstanding writes are always flushed at close() or fsync*() events. This improves the write throughput, especially in case of connections with high latency.
    * Added Windows Client Beta which uses Eldos' Callback File System product instead of Dokan.
    * Re-wrote HDFS interface implementation: Use XtreemFS as replacement for HDFS in your Hadoop setup.
    * libxtreemfs for Java: Access XtreemFS directly from your Java application. See source files org.xtreemfs.common.libxtreemfs.{Client,Volume,FileHandle}.java for the interface.
    * Re-added Vivaldi support and added/improved its functionality:

java/xtreemfs-servers/src/main/java/org/xtreemfs/utils/xtfs_remove_osd.java
/*
 * Copyright (c) 2009-2011 by Paul Seiferth,
 *                            Zuse Institute Berlin
 *
 * Licensed under the BSD License, see LICENSE file for details.
 *
 */
package org.xtreemfs.utils;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.xtreemfs.common.uuids.ServiceUUID;
import org.xtreemfs.common.uuids.UUIDResolver;
import org.xtreemfs.dir.DIRClient;
import org.xtreemfs.foundation.SSLOptions;
import org.xtreemfs.foundation.TimeSync;
import org.xtreemfs.foundation.logging.Logging;
import org.xtreemfs.foundation.logging.Logging.Category;
import org.xtreemfs.foundation.pbrpc.Schemes;
import org.xtreemfs.foundation.pbrpc.client.RPCNIOSocketClient;
import org.xtreemfs.foundation.pbrpc.generatedinterfaces.RPC.Auth;
import org.xtreemfs.foundation.pbrpc.generatedinterfaces.RPC.AuthPassword;
import org.xtreemfs.foundation.pbrpc.generatedinterfaces.RPC.AuthType;
import org.xtreemfs.foundation.pbrpc.generatedinterfaces.RPC.UserCredentials;
import org.xtreemfs.foundation.util.CLIParser;
import org.xtreemfs.foundation.util.CLIParser.CliOption;
import org.xtreemfs.foundation.util.OutputUtils;
import org.xtreemfs.osd.drain.OSDDrain;
import org.xtreemfs.pbrpc.generatedinterfaces.DIR.ServiceSet;
import org.xtreemfs.pbrpc.generatedinterfaces.DIR.ServiceType;
import org.xtreemfs.pbrpc.generatedinterfaces.DIRServiceClient;
import org.xtreemfs.pbrpc.generatedinterfaces.GlobalTypes.PORTS;
import org.xtreemfs.pbrpc.generatedinterfaces.MRCServiceClient;
import org.xtreemfs.pbrpc.generatedinterfaces.OSDServiceClient;

public class xtfs_remove_osd {

    private OSDServiceClient   osd;
    private DIRClient          dir;
    private MRCServiceClient   mrc;
    private RPCNIOSocketClient dirClient;
    private RPCNIOSocketClient osdClient;
    private RPCNIOSocketClient mrcClient;
    private InetSocketAddress  osdAddr;
    private InetSocketAddress  mrcAddr;
    private SSLOptions         sslOptions;
    private String[]           dirAddresses;
    private UUIDResolver       resolver;
    private RPCNIOSocketClient resolverClient;
    private Auth               authHeader;
    private UserCredentials    credentials;
    private String             osdUUIDString;
    private ServiceUUID        osdUUID;

    public static void main(String[] args) {

        Map<String, CliOption> options = null;
        try {
            // parse the call arguments
            options = utils.getDefaultAdminToolOptions(true);
            List<String> arguments = new ArrayList<String>(1);

            CliOption oDir = new CliOption(CliOption.OPTIONTYPE.STRING,
                    "directory service to use (e.g. 'pbrpc://localhost:32638')", "<uri>");
            oDir.urlDefaultPort = PORTS.DIR_PBRPC_PORT_DEFAULT.getNumber();
            oDir.urlDefaultProtocol = Schemes.SCHEME_PBRPC;
            options.put("dir", oDir);
            options.put("s", new CliOption(CliOption.OPTIONTYPE.SWITCH, "shutdown OSD", ""));
            options.put("d", new CliOption(CliOption.OPTIONTYPE.SWITCH, "enbable debug output", ""));
            CLIParser.parseCLI(args, options, arguments);

            // start logging
            if (options.get("d").switchValue) {
                Logging.start(Logging.LEVEL_DEBUG);
            } else {
                Logging.start(Logging.LEVEL_ERROR);
            }

            if (options.get(utils.OPTION_HELP).switchValue || options.get(utils.OPTION_HELP_LONG).switchValue
                    || arguments.size() == 0) {
                usage(options);
                return;
            }

            if (arguments.size() > 1) {
                // print error.
                error("invalid number of arguments", options, true);
            }

            boolean shutdown = options.get("s").switchValue;
            String password = (options.get(utils.OPTION_ADMIN_PASS).stringValue != null) ? options
                    .get(utils.OPTION_ADMIN_PASS).stringValue : "";
            if (password.equals("-")) {
                password = utils.readPassword("Enter admin password: ");
            }

            String[] dirURLs = (options.get("dir").stringValue != null) ? options.get("dir").stringValue
                    .split(",") : null;

            // read default settings for the OSD
            String osdUUID = null;
            if (arguments.get(0).startsWith("uuid:")) {
                osdUUID = arguments.get(0).substring("uuid:".length());
            } else {
                error("There was no UUID for the OSD given!", options);
            }

            SSLOptions sslOptions = null;
            String[] dirAddrs = null;

            // parse security info if protocol is 'https'
            if (dirURLs != null) {
                int i = 0;
                boolean gridSSL = false;
                dirAddrs = new String[dirURLs.length];
                for (String dirURL : dirURLs) {

                    // parse security info if protocol is 'https'
                    if (dirURL.contains(Schemes.SCHEME_PBRPCS + "://")
                            || dirURL.contains(Schemes.SCHEME_PBRPCG + "://") && sslOptions == null) {
                        String serviceCredsFile = options.get(utils.OPTION_USER_CREDS_FILE).stringValue;
                        String serviceCredsPass = options.get(utils.OPTION_USER_CREDS_PASS).stringValue;
                        if(serviceCredsPass != null && serviceCredsPass.equals("-")) {
                        	serviceCredsPass = utils.readPassword("Enter credentials password: ");
                        }
                        String trustedCAsFile = options.get(utils.OPTION_TRUSTSTORE_FILE).stringValue;
                        String trustedCAsPass = options.get(utils.OPTION_TRUSTSTORE_PASS).stringValue;
                        if(trustedCAsPass != null && trustedCAsPass.equals("-")) {
                        	trustedCAsPass = utils.readPassword("Enter trust store password: ");
                        }
                        String sslProtocolString = options.get(utils.OPTION_SSL_PROTOCOL).stringValue;
                        if (dirURL.contains(Schemes.SCHEME_PBRPCG + "://")) {
                            gridSSL = true;
                        }

                        if (serviceCredsFile == null) {
                            System.out.println("SSL requires '-" + utils.OPTION_USER_CREDS_FILE
                                    + "' parameter to be specified");
                            usage(options);
                            System.exit(1);
                        } else if (trustedCAsFile == null) {
                            System.out.println("SSL requires '-" + utils.OPTION_TRUSTSTORE_FILE
                                    + "' parameter to be specified");
                            usage(options);
                            System.exit(1);
                        }

                        // TODO: support custom SSL trust managers
                        try {
                            sslOptions = new SSLOptions(serviceCredsFile, serviceCredsPass,
                                    SSLOptions.PKCS12_CONTAINER, trustedCAsFile, trustedCAsPass,
                                    SSLOptions.JKS_CONTAINER, false, gridSSL, sslProtocolString, null);
                        } catch (Exception e) {
                            System.err.println("unable to get SSL options, because:" + e.getMessage());
                            System.exit(1);
                        }
                    }

                    // add URL to dirAddrs
                    if (dirURL.contains("://")) {
                        // remove Protocol information
                        String[] tmp = dirURL.split("://");
                        // remove possible slash
                        dirAddrs[i++] = tmp[1].replace("/", "");
                    } else {
                        // remove possible slash
                        dirAddrs[i++] = dirURL.replace("/", "");
                    }
                }
            }

            // read default settings
            if (dirURLs == null) {
                try {
                    DefaultDirConfig cfg = new DefaultDirConfig();
                    sslOptions = cfg.getSSLOptions();
                    dirAddrs = cfg.getDirectoryServices();
                } catch (Exception e) {
                    System.err.println("unable to get SSL options, because: " + e.getMessage());
                    System.exit(1);
                }
            }

            xtfs_remove_osd removeOsd = new xtfs_remove_osd(dirAddrs, osdUUID, sslOptions, password);
            removeOsd.initialize();
            removeOsd.drainOSD(shutdown);
            removeOsd.shutdown();

            System.exit(0);

        } catch (Exception e) {

            error(e.getMessage(), options);
        }
    }

    public xtfs_remove_osd(String[] dirAddresses, String osdUUIDString, SSLOptions sslOptions, String password)
            throws Exception {
        try {

            this.sslOptions = sslOptions;
            this.dirAddresses = dirAddresses;
            this.osdUUIDString = osdUUIDString;
            if (password.equals("")) {
                this.authHeader = Auth.newBuilder().setAuthType(AuthType.AUTH_NONE).build();
            } else {
                this.authHeader = Auth.newBuilder().setAuthType(AuthType.AUTH_PASSWORD)
                        .setAuthPasswd(AuthPassword.newBuilder().setPassword(password).build()).build();
            }

            // TODO: use REAL user credentials (this is a SECURITY HOLE)
            this.credentials = UserCredentials.newBuilder().setUsername("root").addGroups("root").build();

        } catch (Exception e) {
            shutdown();
            throw e;
        }
    }

    public void initialize() throws Exception {

        TimeSync.initializeLocal(0);

        // TODO(lukas): support multiple DIRs
        // connect to DIR
        dirClient = new RPCNIOSocketClient(sslOptions, 10000, 5 * 60 * 1000, "xtfs_remove_osd (dir)");
        dirClient.start();
        dirClient.waitForStartup();
        String[] dirAddr = dirAddresses[0].split(":");
        InetSocketAddress dirSocketAddr = new InetSocketAddress(dirAddr[0], Integer.parseInt(dirAddr[1]));
        DIRServiceClient tmp = new DIRServiceClient(dirClient, dirSocketAddr);
        dir = new DIRClient(tmp, new InetSocketAddress[] { dirSocketAddr }, 100, 15 * 1000);

        resolverClient = new RPCNIOSocketClient(sslOptions, 10000, 5 * 60 * 1000, "xtfs_remove_osd (resolver)");
        resolverClient.start();
        resolverClient.waitForStartup();
        this.resolver = UUIDResolver.startNonSingelton(dir, 1000, 10 * 10 * 1000);

        // create OSD client
        osdUUID = new ServiceUUID(osdUUIDString, resolver);
        osdUUID.resolve();
        osdAddr = osdUUID.getAddress();

        osdClient = new RPCNIOSocketClient(sslOptions, 10000, 5 * 60 * 1000, "xtfs_remove_osd (osd)");
        osdClient.start();
        osdClient.waitForStartup();
        osd = new OSDServiceClient(osdClient, osdAddr);

        // create MRC client
        ServiceSet sSet = null;
        try {
            sSet = dir.xtreemfs_service_get_by_type(null, authHeader, credentials,
                    ServiceType.SERVICE_TYPE_MRC);
        } catch (IOException ioe) {
            Logging.logMessage(Logging.LEVEL_WARN, Category.proc, new Object(),
                    OutputUtils.stackTraceToString(ioe));
            throw ioe;
        }

        mrcClient = new RPCNIOSocketClient(sslOptions, 100000, 5 * 60 * 10000, "xtfs_remove_osd (mrc)");
        mrcClient.start();
        mrcClient.waitForStartup();

        if (sSet.getServicesCount() == 0) {
            throw new IOException("No MRC is currently registred at DIR");
        }

        String mrcUUID = sSet.getServices(0).getUuid();
        ServiceUUID UUIDService = new ServiceUUID(mrcUUID, resolver);
        UUIDService.resolve();
        mrcAddr = UUIDService.getAddress();

        mrc = new MRCServiceClient(mrcClient, mrcAddr);

    }

    public void shutdown() {
        try {
            resolverClient.shutdown();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * Removes (drain) an OSD.
     * 
     * @throws Exception
     */
    public void drainOSD(boolean shutdown) throws Exception {
        OSDDrain osdDrain = new OSDDrain(dir, osd, mrc, osdUUID, authHeader, credentials, resolver);
        osdDrain.drain(shutdown);
    }

    /**
     * Prints the error <code>message</code> and delegates to usage() if "printUsage" is true.
     * 
     * @param message
     *            The error message
     * @param options
     *            The CLI Options.
     * @param printUsage
     *            True if usage should be printed. False otherwise.
     * 
     */
    private static void error(String message, Map<String, CliOption> options, boolean printUsage) {
        System.err.println(message);

        if (printUsage) {
            System.out.println();
            usage(options);
        }
        System.exit(1);
    }

    /**
     * Prints the error <code>message</code> and delegates to usage().
     * 
     * @param message
     *            The error message
     * @param options
     *            The CLI Options.
     */
    private static void error(String message, Map<String, CliOption> options) {
        error(message, options, false);
    }

    public static void usage(Map<String, CliOption> options) {

        System.out.println("usage: xtfs_remove_osd [options] uuid:<osd_uuid>\n");
        System.out.println("  " + "<osd_uuid> the unique identifier of the OSD to be removed\n");
        System.out.println("  " + "options:");

        utils.printOptions(options);
    }
}
